compiler/src/iree/compiler/Dialect/LinalgExt/Utils/MatchUtils.cpp

```diff
@@ -0,0 +1,316 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/LinalgExt/Utils/MatchUtils.h"
+
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SetOperations.h"
+#include "llvm/ADT/SmallBitVector.h"
+#include "llvm/ADT/SmallVector.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Complex/IR/Complex.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
+#include "mlir/IR/DialectImplementation.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/TypeUtilities.h"
+
+// This is based on the upstream implementation of the
+// Linalg::ContractionOpInterface.
+
+namespace mlir::iree_compiler::IREE::LinalgExt {
+
+namespace {
+constexpr auto par = utils::IteratorType::parallel;
+constexpr auto red = utils::IteratorType::reduction;
+} // namespace
+
+namespace detail {
+enum class MatchContractionResult {
+  Success = 0,
+  NotLinalgOp,
+  WrongNumOperands,
+  NoReduction,
+  NotProjectedPermutations,
+  NotAddMul
+};
+}
+
+//===----------------------------------------------------------------------===//
+// ScaledContractionOpInterface implementation
+//===----------------------------------------------------------------------===//
+
+/// If the value is defined by a chain of unary side-effect free, go up the
+/// use-def chain until the first value that isn't defined by such an op.
+// TODO: relax to multi-operands with constants, which are technically unary ops
+// as needed (e.g. add5).
+static Value getSourceSkipUnary(Value value) {
+  Operation *op = value.getDefiningOp();
+  while (op && op->getNumOperands() == 1) {
+    auto iface = dyn_cast<MemoryEffectOpInterface>(op);
+    if (!iface || !iface.hasNoEffect()) {
+      break;
+    }
+    value = op->getOperand(0);
+    op = value.getDefiningOp();
+  }
+  return value;
+}
+
+/// Returns true if the two operations are ones supported by scaled mfmas.
+static bool isValidScaledMmaPair(Operation *add, Operation *mul) {
+  SmallVector<std::pair<std::string, std::string>> validPairs = {
```

Comment:
Use StringRef or StringLiteral to avoid dynamic memory allocations. These can be `constexpr` and in an array (we don't ever have to resize this).